### ISSUE TITLE
Importing database is extremely slow

### ISSUE THREAD

**Issue by generalredneck** [lando/lando#456]
**Tell us about your setup**

What is your lando version and operating system?
lando v3.0.0-beta.20
docker 17.09.0-ce
ubuntu Gnome 16.04.3 LTS 64-bit

**Tell us about your `.lando.yml`**

```yml
name: accme                                                                                                                                                                                                                                   
recipe: drupal7                                                                 
config:                                                                         
  webroot: ./web                                                                
  php: '5.6'
```

As a lando user, I would like databases to not take so long to import so that I can get more work done.

So I rebuild environments a lot. We are evaluating Lando as something to replace DrupalVM and I happened to notice that database imports seemed to take so much longer so I did a bench mark of running drush commands using shell aliases, the standard lando db-import, and finally the same drush shell alias used in the first test against Drupal VM set up with NFS.

For reference sake... `drush fresh` in my case runs the following:

```
$options['shell-aliases']['fresh'] = implode("\n", array(                       
  "!echo '\nDropping the database...'",                                         
  "drush sql-drop -y",                                                          
  "echo '\nImporting reference database...'",                                   
  "gunzip < ../reference/accme.sql.gz | drush sql-cli",                         
));
```

The following are the lowest benchmark on the same database I was importing which was 79MB...

```
$ time lando drush fresh

Dropping the database...
Do you really want to drop all tables in the database drupal7? (y/n): y

Importing reference database...

real    5m5.393s
user    0m0.724s
sys    0m0.080s
```

```
$ time lando db-import ../reference/accme-live-2017-11-01-08-13-04.sql.gz 
Preparing to import /app/web/../reference/accme-live-2017-11-01-08-13-04.sql.gz into drupal7 on database:3306 as drupal7...
Destroying all current tables in drupal7... 
NOTE: See the --no-wipe flag to avoid this step!
Dropping [...]
Gunzipped file detected!
Importing /app/web/../reference/accme-live-2017-11-01-08-13-04.sql.gz...
78.2MiB 0:04:56 [ 270KiB/s] [================================>] 100%            
Import completed with status code 0

real    5m19.920s
user    0m1.480s
sys    0m0.140s
```

``` 
time drush @accme.local fresh

Dropping the database...
Do you really want to drop all tables in the database drupal? (y/n): y

Importing reference database...

real    0m58.171s
user    0m0.072s
sys    0m0.128s
```

Any ideas? I know this is comparing a VM vs Docker solution... but is there any way we can speed this up from a 5 minutes? As you can see Drupal VM is only taking about 1 minute... and on larger databases this number seems to scale.

**Comment by pirog** [comment]
@generalredneck i'm not exactly sure why this would be the case as there are many things that could impact the speed. my *best guess* is this is just the difference between docker and a dedicated vm. if that's the case then i think we just have to wait for docker itself to make improvements in this space and there isnt much we can do on the lando front beyond bugging them to make those improvements. 

every local dev solution has pros and cons. generally i think the pros you get from something like lando outweigh the cons but that is sadly not a universal truth and you've got to weight the pros and cons that make sense to you. that said, and regardless of your ultimate local dev choice, it might be worth examining why your workflow requires so many database rebuilds. certainly there could be completely legit reasons why you need to do that but generally its worth trying to limit these kinds of operations because, as you point out, they can be costly.

im going to icebox this for now and hope that we can get more user feedback

**Comment by generalredneck** [comment]
What I suggest we do is first determine if it's network or filesystem related. I'm not willing to settle with "just the difference between docker and a dedicated vm"

Googling the issue there are a ton of people that get disk performance to be equal of that of the native disk using default docker as long as the filesystems are mounted correctly. 

Both are pulling off of a file share... I just don't know the difference and if db-import is actually pushing over the internal network between containers while readinf from a share... or what... on the VM I'm using nfs to share the file to the VM so that's the main factor there...

So as you said we will icebox it until someone with better docker experience chimes in or I learn more.

**Comment by ccharlton** [comment]
For reference, one of the db's I sync with is 2GB and takes over 22 minutes using 'lando pull' (which is convenient but bad for db import speed), but using the `db-import` on an archive takes about 7 minutes (but I have to wait for an archive to be generated if I need the freshest data, on Pantheon I use the midnight auto-backups since they're already available.

**Comment by generalredneck** [comment]
@ccharlton but what I'm saying is that db-import shouldn't even be taking THAT long because I use the same method as you do... We rarely need the freshest data, but we *do* need to see how features will build since at least the last release... so the nightly database backups are good candidates for this... Thus why I was comparing `db-import` with `drush sql-drop && gunzip < db.sql.gz | drush sqlc` on both methods of development.

**Comment by generalredneck** [comment]
FYI, after looking at #561 and the answer there... there was some debate internally here at Four Kitchens... and one of the suggestions was that it was the file sharing method that was causing the slow down... I feel like I effectively removed that as the possibility by copying the sql.gz file to the root file system of the container (in my case /var/www). So I did that and sshed in and did the following

```
www-data@80975921eef5:/app/web$ time gunzip < /var/www/accme-live-2017-11-01-08-13-04.sql.gz | drush sqlc

real	5m20.341s
user	0m8.148s
sys	0m1.068s
```
Furthermore I wanted to rule out it's a "networking" problem so I sshed into the database container, copied the file to /var/www, and tried again.

```
www-data@1332c7483000:/app$ time gunzip < ~/accme-live-2017-11-01-08-13-04.sql.gz | mysql --user=drupal7 --password=drupal7 --database=drupal7 --host=localhost --port=3306
mysql: [Warning] Using a password on the command line interface can be insecure.

real	5m21.780s
user	0m9.116s
sys	0m0.760s
```

Next to verify that read write times to the disk didn't affect this... I did some writing to the disk using dd both in the docker container (shared and unshared folder) and in a drupal-vm shared folder (using nfs).

There was a big variance in the write speeds but the variance was consistent between all of the above.  
Example if I was getting ~110 MB/s in docker, I would get ~110 MB/s in drupal-vm... I got between 170 MB/s and 60 MB/s on both systems. I ran the following. I didn't disable disk cache FYI.

`dd if=/dev/zero of=/var/www/testfile bs=1G count=1 oflag=direct`
`dd if=/dev/zero of=/app/testfile bs=1G count=1 oflag=direct`
`dd if=/dev/zero of=/var/www/drupalvm/web/testfile bs=1G count=1 oflag=direct`

So my thought is there must be something up with the mysql implementation but I would like for someone to take a look at what I did above and verify that what I did makes sense and does in fact rule out IO latency.

The other thought may be around processing power... but I'm not sure...

From the database container I ran `lscpu` just to see how it matched up with my VM

```
www-data@1332c7483000:/app$ lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    2
Core(s) per socket:    2
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 142
Model name:            Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
Stepping:              9
CPU MHz:               603.399
CPU max MHz:           3500.0000
CPU min MHz:           400.0000
BogoMIPS:              5808.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              4096K
NUMA node0 CPU(s):     0-3
```
VS

```
vagrant@accme:/var/www/drupalvm/web$ lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                1
On-line CPU(s) list:   0
Thread(s) per core:    1
Core(s) per socket:    1
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 142
Model name:            Intel(R) Core(TM) i7-7500U CPU @ 2.70GHz
Stepping:              9
CPU MHz:               2904.000
BogoMIPS:              5808.00
Hypervisor vendor:     KVM
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              4096K
NUMA node0 CPU(s):     0
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 syscall nx rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc pni pclmulqdq monitor ssse3 cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx rdrand hypervisor lahf_lm abm 3dnowprefetch rdseed clflushopt
```

I've not learned how to add CPU resources to docker... and therefore I'm not sure I can test if increasing that will make a difference...

Please let me know if I'm way off base here, but I have a hard time believing this is a specifically "docker" problem, and not something we can improve upon.

**Comment by uberhacker** [comment]
@dustinleblanc: This has been a long-standing issue for Pantheon downloads via `lando pull`.  I'm not convinced this is a Lando only issue as Pantheon downloads are considerably slower than downloads from other projects including Acquia Cloud and Platform.sh.  FWIW, downloading from the Pantheon dashboard is very quick usually which leads me to believe it's not necessarily a bandwidth issue.

**Comment by dustinleblanc** [comment]
Yep @pirog and I have talked about slow downloads via Pantheon and thought about ways to mitigate. The story above seems to indicate that pulls of gzipped SQL files in our containers (Lando MySQL) are also slow

**Comment by zuernBernhard** [comment]
FYI my performance issues with landy mysql on macos went away by upgrading "docker for mac" from the "stable" release to "edge".

**Comment by generalredneck** [comment]
@dustinleblanc,
Sorry for some reason I missed the last few correspondences here... I would be interested in trying out a new container. Also... let me test again on the latest Docker CE that I have for linux... 

@uberhacker I totally get that `lando pull` is slow... and I've mitigated that by using latest nightly backups even on my other implementations. As Dustin said... totally a separate issue. It's mostly about `lando db-import` or straight up drush calls that I had issues with.

**Comment by generalredneck** [comment]
@dustinleblanc 
For what it's worth, I got a pretty close facsimile to the drush command, but not verbetium with something like:
```
docker pull mysql:latest
docker run --name some-mysql -e MYSQL_ROOT_PASSWORD=my-secret-pw -d mysql:latest
docker exec -it some-mysql sh -c 'mysql -e "CREATE DATABASE test" -h"localhost" -P"3306" -uroot -p"my-secret-pw"'
docker cp ../pri.org-new/reference/db.sql.gz some-mysql:/tmp
time docker exec -it some-mysql sh -c ' gunzip < /tmp/db.sql.gz | mysql -h"localhost" -P"3306" -uroot -p"my-secret-pw" test'
```
It's safe to say that running it in that container is just as slow as my lando implementation. So it's very possible this is a "docker" thing... Or at least a mysql container thing... and I'm glad yall were able to come up with a way to point that out. The next question is, where do we go from here?

Just icing below... I still feel that using DrupalVM makes the most sense for migrations and what not because of this... on the other hand... the ease of use, quick provisioning, and lack of failure of lando makes it a great choice for every day site building (especially for less devopsy people)... You just literally lose seconds on every task you do all day long because of slow database operations. 

All that said, I like lando, alot... I just wish there was more I could do about :point_up:

**Comment by dalin-** [comment]
Is it possibly a MySQL configuration thing?  The stock `my.cnf` is known to be be very slow as it prioritizes absolute data integrity over anything else.  Which is not at all useful for a local install.  It's been several years since I played with it, but I remember a few settings making *huge* improvements:

```ini
[mysqld]
tmp_table_size            = 96M
max_heap_table_size       = 96M
open_files_limit          = 24000
table_definition_cache    = 4096
table_open_cache          = 256
sort_buffer_size          = 64K
read_buffer_size          = 256K
read_rnd_buffer_size      = 256K
key-buffer-size           = 32M
myisam-recover            = FORCE,BACKUP
log-warnings                   = 2
# Not certain about this one in the context of a containerized setup:
skip-networking

# This group is where you get the biggest bang for the buck:
innodb-flush-method            = O_DIRECT
innodb-flush-log-at-trx-commit = 0
innodb-file-per-table          = 1
innodb-buffer-pool-size        = 256M
# Setting these two has to be done before MySQL is started for the first time:
innodb-log-files-in-group      = 2
innodb-log-file-size           = 64M

# This one may be specific to Percona, and might throw errors on MariaDB:
performance_schema        = OFF

**Comment by pirog** [comment]
@dalin- move of the recipes use a custom mysql conf
https://github.com/lando/lando/blob/master/plugins/lando-recipes/recipes/pantheon/mysql/my.cnf

**Comment by osopolar** [comment]
Maybe you can speed it up a bit by disabling the [innodb doublewrite buffer](https://dev.mysql.com/doc/refman/5.7/en/innodb-doublewrite-buffer.html), adding `innodb_doublewrite = 0` to your my.cnf file.

**Comment by hkirsman** [comment]
What's this https://gist.github.com/mkrakauer-rio/e7d9de75f5ac680e790365748ca188a4 . Could not enable it. Or at least I think I didn't.

More read here: https://dzone.com/articles/docker-for-mac-performance-tweaks-1

I think this is for for older Docker?

Or what's this:
https://github.com/docker/for-linux/issues/247 Only for Linux?

**Comment by dustinleblanc** [comment]
@hkirsman The database flushing thing is interesting for sure, might have to come back to that for review at a later date. Not sure we want to decide on speed over data integrity for ALL users, even if it is a fairly good assumption.

**Comment by wturrell** [comment]
Where can I find some documentation on how to customise Lando's my.cnf? 

I see there's a file in _~/.lando/services/config/drupal8/mysql/_, does that get copied somewhere else?

**Comment by jimafisk** [comment]
@wturrell here is some documentation on using a custom my.cnf override: https://docs.devwithlando.io/tutorials/mysql.html#using-a-custom-mysql-config-file

I haven't been able to get it to work though, not sure what I'm missing.

**Comment by ccharlton** [comment]
I prefer to use `lando db-import DBFILE.sql.gz` instead of `lando pull`; it's 1000x times faster. Quote me on that, if you want. ;)

**Comment by Fonata** [comment]
The following is for Lando end users and continuous integration, not for Lando itself. If you are ok with re-importing the database after every Lando start, you can use something like the following in your ``.lando.yml`` to force the database to be stored in memory:
````yml
services:
  database:
    overrides:
      volumes:
        - type: tmpfs
          target: /bitnami/mysql/data
````

**Comment by kirantejp** [comment]
I am using lando (v3.0.0.rc-14) along with docker CE (latest version) for my Drupal 7 site. I am trying to import database of file size uncomressed(~4GB) & compressed (~900 MB) using following command:

lando db-import db-name db-filename.sql
lando db-import db-name db-filename.sql.gz

But, this is not helping. Usually, this db-import is taking more than 24 hours to complete. Is this a problem with my version or my settings in .lando.yml file?

I'm using Ubuntu (18.04) OS and my hardware config is 16 GB RAM.

**Comment by dalin-** [comment]
I've not worked with a DB that big, the largest one that I work with is probably about half of that. And that imports in about 5–15mins (I can't remember exactly). I also don't use `lando db-import` but rather `lando pull` which is potentially faster since it doesn't need to read/write to the disk at the same time. 

My guess is that you don't have enough RAM allocated to Docker and thus it's paging out to disk.

**Comment by ccharlton** [comment]
For additional reference, a 2 GB db import via 'lando db-import' usually takes my machine 8-13 minutes to complete depending on what else I got going on.

If your 4 GB db is taking >1 hour to import, you may want to do a machine/settings gut-check because that seems abnormally too long.

**Comment by kirantejp** [comment]
Hi @dalin- After your suggestion, I increased the memory size for mysqld container to 5GB with following command:

`docker run -e MYSQL_ROOT_PASSWORD=mypassword -d --memory 5g mysql`

But, still when I checked CPU & Memory usage, that mysqld container is still using 3.5% of memory only. Also, I switched to lando stable version (`lando-v3.0.0-beta.47`). Even after these changes, `lando db-import` is still running for > 6 hours. 

So, now @ccharlton , should I check my machine hardware settings or lando/docker settings. Please suggest.

**Comment by kirantejp** [comment]
Any update on this please??

**Comment by ccharlton** [comment]
To answer the `innodb_buffer_pool_size` portion: I do have that bumped up for my larger Lando db's, and other settings.

(I'm now using the new RC14+ Lando version, up from Beta.47 like other folks here)

**Comment by generalredneck** [comment]
just to point out, due to some people commenting on syncing from pantheon, someone labeled this as "pantheon". this is not about the pantheon sync, but the speed of which sql imports into mysql on a lando container. This is very much a performance based issue and though we have seen improvements in this realm, it's very much still a thing compared to other development platforms such as drupalvm and some local server installations using apache, mysql, php. It's clear that there will be no further improvements in this area besides if something upstream comes in, something accidentally makes this better, or there are clear small improvements that can be made and tested versus the over-sweeping statement "Mysql in lando is slow." At that point a new issue can be opened with those pointed improvements and this may be referenced if needed.

**Comment by mchelen** [comment]
Related work going on for this here: https://github.com/lando/lando/issues/2402

**Comment by pirog** [comment]
As indicated by @generalredneck in https://github.com/lando/lando/issues/456#issuecomment-500240140

> this is not about the pantheon sync, but the speed of which sql imports into mysql on a lando container.
